simpleble/src/backends/common/PeripheralBase.h

```diff
@@ -1,13 +1,85 @@
-#elif SIMPLEBLE_BACKEND_ANDROID == 1
-#include "../android/PeripheralAndroid.h"
-#endif
+#include <functional>
+#include <map>
+#include <memory>
+#include <vector>
+
+#include <simpleble/Types.h>
+
+namespace SimpleBLE {
+
+class ServiceBase;
+
+/**
+ * Abstract base class for Bluetooth adapter implementations.
+ *
+ * Each backend must implement this class.
+ *
+ * Notes for implementers: ...
+ */
+class PeripheralBase {
+  public:
+    virtual ~PeripheralBase() = default;
+
+    virtual void* underlying() const = 0;
+
+    virtual std::string identifier() = 0;
+    virtual BluetoothAddress address() = 0;
+    virtual BluetoothAddressType address_type() = 0;
+    virtual int16_t rssi() = 0;
+
+    /**
+     * @brief Provides the advertised transmit power in dBm.
+     *
+     * @note If the field has not been advertised by the peripheral,
+     *       the returned value will be -32768.
+     */
+    virtual int16_t tx_power() = 0;
+    virtual uint16_t mtu() = 0;
+
+    virtual void connect() = 0;
+    virtual void disconnect() = 0;
+    virtual bool is_connected() = 0;
+    virtual bool is_connectable() = 0;
+    virtual bool is_paired() = 0;
+    virtual void unpair() = 0;
+
+    /**
+     * Available services (if the peripheral is connected).
+     */
+    virtual std::vector<std::shared_ptr<ServiceBase>> available_services() = 0;
+
+    /**
+     * Advertised services (if the peripheral is not connected).
+     */
+    virtual std::vector<std::shared_ptr<ServiceBase>> advertised_services() = 0;
+
+    virtual std::map<uint16_t, ByteArray> manufacturer_data() = 0;
+
+    // clang-format off
+    /* These methods call the virtual counterparts ONLY if the device is connected,
+       otherwise they throw Exception::NotConnected */
+
+    /* These methods are called by the frontend ONLY when the device is connected.
+    */
+    // clang-format off
+    virtual ByteArray read_inner(BluetoothUUID const& service, BluetoothUUID const& characteristic) = 0;
+    virtual void write_request_inner(BluetoothUUID const& service, BluetoothUUID const& characteristic, ByteArray const& data) = 0;
+    virtual void write_command_inner(BluetoothUUID const& service, BluetoothUUID const& characteristic, ByteArray const& data) = 0;
+    virtual void notify_inner(BluetoothUUID const& service, BluetoothUUID const& characteristic, std::function<void(ByteArray payload)> callback) = 0;
+    virtual void indicate_inner(BluetoothUUID const& service, BluetoothUUID const& characteristic, std::function<void(ByteArray payload)> callback) = 0;
+    virtual void unsubscribe_inner(BluetoothUUID const& service, BluetoothUUID const& characteristic) = 0;
+
+    virtual ByteArray read_inner(BluetoothUUID const& service, BluetoothUUID const& characteristic, BluetoothUUID const& descriptor) = 0;
+    virtual void write_inner(BluetoothUUID const& service, BluetoothUUID const& characteristic, BluetoothUUID const& descriptor, ByteArray const& data) = 0;
```

Comment:
Can we skip adding the `_inner` suffix? It's sort of understandable if you're calling the `internal` pointer that you're referring to an inner method.

Reply:
I put the _inner to indicate that it is not just that the frontend method calls the implementation method, but it also checks whether it is connected, so the _inner methods are only called on a connected peripheral. If that indication is not important, I can change the name back.